addons/mail/models/mail_blacklist.py

```diff
@@ -0,0 +1,84 @@
+    active = fields.Boolean(default=True, track_visibility='onchange')
+
+    _sql_constraints = [
+        ('unique_email', 'unique (email)', 'Email address already exists!')
+    ]
+
+    @api.model
+    def create(self, values):
+        """ To avoid crash during import due to unique email, return the existing record if any """
+        rec = self.search([('email', '=ilike', values['email'])])
+        if len(rec) == 1:
+            return rec
+        return super(MailBlackList, self).create(values)
+
+    def _add(self, email):
+        blacklist = self.env["mail.blacklist"].sudo()
+        record = blacklist.with_context(active_test=False).search([('email', '=', email)])
+        if len(record) > 0:
+            record.active = True
+        else:
+            record = self.create({'email': email})
+        return record
+
+    def _remove(self, email):
+        blacklist = self.env["mail.blacklist"].sudo()
+        record = blacklist.with_context(active_test=False).search([('email', '=', email)])
+        if len(record) > 0:
+            record.active = False
+        else:
+            raise UserError(_('This email does not belong to the blacklist.'))
+
+
+class MailBlackListMixin(models.AbstractModel):
+    """ Mixin that is inherited by all model with opt out.
+        USAGE : the field '_primary_email' must be overridden in the model that inherit the mixin
+        and must contain the email field of the model.
+        """
+    _name = 'mail.blacklist.mixin'
+    _description = 'Mail Blacklist mixin'
+    _primary_email = ['email']
+
+    # Note : is_blacklisted sould only be used for display. As the compute is not depending on the blacklist,
+    # once read, it won't be re-computed again if the blacklist is modified in the same request.
+    is_blacklisted = fields.Boolean(string='Blacklist', compute="_compute_is_blacklisted", store=False,
+        search="_search_is_blacklisted",
+        help="If the email address is on the blacklist, the contact won't receive mass mailing anymore, from any list",
+        groups="base.group_user")
+
+    @api.model
+    def _search_is_blacklisted(self, operator, value):
+        [email_field] = self._primary_email
+        join_condition = 'INNER' if value else 'RIGHT'
+        where_clause = '' if value else ' where b.id is null'
+        blacklisted_sql = '''SELECT a.id FROM mail_blacklist b 
+            %s JOIN %s a 
+            ON b.email = a.%s AND b.active = True%s''' % (join_condition, self._table, email_field, where_clause)
+        self._cr.execute(blacklisted_sql)
+        res = self._cr.fetchall()
+        if not res:
+            return [(0, '=', 1)]
+        return [('id', 'in', [r[0] for r in res])]
+
+    @api.depends(lambda self: self._primary_email)
+    def _compute_is_blacklisted(self):
+        [email_field] = self._primary_email
+        blacklist = self.env['mail.blacklist'].sudo().search([('email', 'in', [record[email_field] for record in self])]).mapped('email')
```

Comment:
1/ rather than use `sudo` here, you can use `compute_sudo=True`on field definition directly
2/ `[record[email_field] for record in self]` can be replaced by `self.mapped(email_field)`. There will be no duplicate value in the list.